compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -105,10 +105,9 @@ object TypeTestsCasts {
       debug.println("P1 : " + P1.show)
       debug.println("X : " + X.show)
 
-      P1 <:< X       // constraint P1
+      P1 <:< X
```

Comment:
I don't undestand what is going on here, the documentation above still says "constrain `Xs` with `pre.F[Xs] <:< X`", but this is not true in general since no constraint will be added if `P1` is not a subtype of `X`. The documentation need to explain why this is safe.

Reply:
That is expected if `P1` is not a subtype of `X`. It just tries to infer type parameters of `P1` from `X` if the value `x` indeed has the outer type `pre.F[_]`. Then it goes on to check if `P1 <: P`, which means the type arguments in `P` are trivial, thus no runtime checks are needed for them. Therefore, no warnings in such cases.

I'll refine the doc.